### PR TITLE
chore: Add gRPC validation for EscrowAddress

### PR DESCRIPTION
**Pull request by bznein**
## Description

Added gRPC validation for EscrowAddress in apps/transfer/keeper

ref: #4140 

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [x] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

## Summary by CodeRabbit

- **New Features**
	- Improved validation logic for gRPC request parameters in the transfer service.
- **Bug Fixes**
	- Enhanced error handling for non-existent channels in transfer queries.
- **Tests**
	- Added new test cases for transfer service to cover scenarios with invalid channel identifiers.

### CODE REVIEW

File: modules/apps/transfer/keeper/grpc_query_test.go [hunk 1]
```diff
@@ -245,12 +245,24 @@ func (suite *KeeperTestSuite) TestEscrowAddress() {
...
+			},
+			false,
+		},
 	}
 
 	for _, tc := range testCases {
 		tc := tc
 		suite.Run(fmt.Sprintf("Case %s", tc.msg), func() {
 			suite.SetupTest() // reset
+			path := ibctesting.NewTransferPath(suite.chainA, suite.chainB)
```
Comment (bznein) [inline modules/apps/transfer/keeper/grpc_query_test.go:264]: I added this part to enable creation of the channel, otherwise the test for the "happy path" would fail.

File: modules/apps/transfer/keeper/grpc_query.go [hunk 2]
```diff
@@ -142,3 +157,15 @@ func (k Keeper) TotalEscrowForDenom(c context.Context, req *types.QueryTotalEscr
 		Amount: amount,
 	}, nil
 }
+
+func validategRPCRequest(portID, channelID string) error {
```
Comment (bznein) [inline modules/apps/transfer/keeper/grpc_query.go:161]: This function is copy-pasted from here: https://github.com/cosmos/ibc-go/blob/1323a733d5b0d83478f1362ebc64cb1e42fa8caa/modules/core/04-channel/keeper/grpc_query.go#L641-L651

we could technically export that one instead but since it's internal logic I'm not sure that would be better. Open to change it in that case.

**Comment by chatton** [comment]
E2E failure looks unrelated, we can re-run it :+1:

File: modules/apps/transfer/keeper/grpc_query.go [hunk 3]
```diff
@@ -15,6 +15,8 @@ import (
 	"github.com/cosmos/cosmos-sdk/types/query"
 
 	"github.com/cosmos/ibc-go/v8/modules/apps/transfer/types"
+	errors "github.com/cosmos/ibc-go/v8/modules/core/04-channel/types"
```
Comment (chatton) [inline modules/apps/transfer/keeper/grpc_query.go:18]: we typically use `channeltypes` as the alias here, the errors are just in the package along with the other types.

File: modules/apps/transfer/types/expected_keepers.go [hunk 4]
```diff
@@ -39,6 +39,8 @@ type ChannelKeeper interface {
 	GetChannel(ctx sdk.Context, srcPort, srcChan string) (channel channeltypes.Channel, found bool)
 	GetNextSequenceSend(ctx sdk.Context, portID, channelID string) (uint64, bool)
 	GetAllChannelsWithPortPrefix(ctx sdk.Context, portPrefix string) []channeltypes.IdentifiedChannel
+
```
Comment (chatton) [inline modules/apps/transfer/types/expected_keepers.go:42]: [nit] can we axe this whitespace :sweat_smile:

**Review by chatton — COMMENTED** [review]
Looks great! I think we can just change the alias to be more consistent then we should be good to go! Great work :muscle:

File: modules/apps/transfer/types/expected_keepers.go (hunk 4 above)
Comment (bznein) [reply modules/apps/transfer/types/expected_keepers.go]: Done

File: modules/apps/transfer/keeper/grpc_query.go (hunk 3 above)
Comment (bznein) [reply modules/apps/transfer/keeper/grpc_query.go]: Done!

**Review by chatton — APPROVED** [review]
LGTM thank you @bznein !

File: modules/apps/transfer/keeper/grpc_query.go (hunk 2 above)
Comment (bznein) [reply modules/apps/transfer/keeper/grpc_query.go:161]: Yeah it could be a nice clean-up after I complete the issue :)

File: modules/apps/transfer/keeper/grpc_query.go [hunk 5]
```diff
@@ -112,13 +114,26 @@ func (k Keeper) DenomHash(c context.Context, req *types.QueryDenomHashRequest) (
...
-func (Keeper) EscrowAddress(c context.Context, req *types.QueryEscrowAddressRequest) (*types.QueryEscrowAddressResponse, error) {
+func (k Keeper) EscrowAddress(c context.Context, req *types.QueryEscrowAddressRequest) (*types.QueryEscrowAddressResponse, error) {
 	if req == nil {
 		return nil, status.Error(codes.InvalidArgument, "empty request")
 	}
 
 	addr := types.GetEscrowAddress(req.PortId, req.ChannelId)
 
+	if err := validategRPCRequest(req.PortId, req.ChannelId); err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (bznein) [reply modules/apps/transfer/keeper/grpc_query.go:125]: Good point, added!

File: modules/apps/transfer/keeper/grpc_query.go [hunk 6]
```diff
@@ -112,13 +114,26 @@ func (k Keeper) DenomHash(c context.Context, req *types.QueryDenomHashRequest) (
...
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+	found := k.channelKeeper.HasChannel(ctx, req.PortId, req.ChannelId)
+	if !found {
```
Comment (damiannolan) [inline modules/apps/transfer/keeper/grpc_query.go:130]: This could be condensed but I don't really feel strongly about it...

```suggestion
	if !k.channelKeeper.HasChannel(ctx, req.PortId, req.ChannelId) {
```

File: modules/apps/transfer/keeper/grpc_query.go [hunk 7]
```diff
@@ -112,13 +114,26 @@ func (k Keeper) DenomHash(c context.Context, req *types.QueryDenomHashRequest) (
...
+		return nil, status.Error(
+			codes.NotFound,
+			errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "port-id: %s, channel-id %s", req.PortId, req.ChannelId).Error(),
```
Comment (damiannolan) [inline modules/apps/transfer/keeper/grpc_query.go:133]: mega formatting nit: see errors msgs in handshake.go 04-channel for example

```suggestion
			errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "port ID (%s) channel ID (%s)", req.PortId, req.ChannelId).Error(),
```

Comment (damiannolan) [reply modules/apps/transfer/keeper/grpc_query.go:133]: to be fair there's probably inconsistency in places, but I think the suggested might be mostly standard 😅

**Review by damiannolan — APPROVED** [review]
Nice one! Thanks for picking this up @bznein 

All I left was nits, lgtm!

File: modules/apps/transfer/keeper/grpc_query_test.go [hunk 8]
```diff
@@ -245,12 +245,34 @@ func (suite *KeeperTestSuite) TestEscrowAddress() {
...
+			"failure - channel not found due to invalid channelID",
+			func() {
+				req = &types.QueryEscrowAddressRequest{
+					PortId:    ibctesting.TransferPort,
+					ChannelId: ibctesting.InvalidID,
+				}
+			},
+			false,
+		},
 	}
```
Comment (crodriguezvega) [inline modules/apps/transfer/keeper/grpc_query_test.go:268]: Should we add a test where both port ID and channel ID are valid, but no channel is found for those parameters? Because the tests will fail on `validategRPCRequest` because either port ID or channel ID is not valid, but we don't have a test case that fails if `HasChannel` returns false, right?

**Review by crodriguezvega — COMMENTED** [review]
Thanks, @bznein. I left a comment because maybe there's one more test case we can add?

File: modules/apps/transfer/keeper/grpc_query_test.go [hunk 9]
```diff
@@ -245,12 +245,34 @@ func (suite *KeeperTestSuite) TestEscrowAddress() {
 			},
 			true,
 		},
+		{
+			"failure - channel not found due to invalid portID",
```
Comment (crodriguezvega) [inline modules/apps/transfer/keeper/grpc_query_test.go:249]: I think I would put it like this, because the code doesn't actually try to find the channel, it fails before that, right?

```suggestion
			"failure - invalid portID",
```

File: modules/apps/transfer/keeper/grpc_query_test.go (hunk 8 above)
Comment (bznein) [reply modules/apps/transfer/keeper/grpc_query_test.go:268]: I might have misunderstood it :) invalidID is equal to IDIsInvalid (https://github.com/cosmos/ibc-go/blob/2555a7c504a904064d659e4c1a3a74000887f73d/testing/values.go#L37) but when we check for its validity in validategRPCRequest (https://github.com/cosmos/ibc-go/blob/main/modules/core/24-host/validate.go#L39-L60) I don't think this value will cause that function to return false? The ID is not blank, doesn't contain slashes, it's within the range for length, and satisfies the regex: https://github.com/cosmos/ibc-go/blob/main/modules/core/24-host/validate.go#L28

I just assumed that "invalidID" meant something like "an ID that would neer match an existing port/channel ID" but maybe I was wrong :)

Comment (bznein) [reply modules/apps/transfer/keeper/grpc_query_test.go:268]: To verify this I commented out the check for `HasChannel` and ran the test again, and got this failure:

```
--- FAIL: TestKeeperTestSuite (55.03s)
    --- FAIL: TestKeeperTestSuite/TestEscrowAddress (1.59s)
        --- FAIL: TestKeeperTestSuite/TestEscrowAddress/Case_failure_-_channel_not_found_due_to_invalid_portID (0.40s)
            grpc_query_test.go:287: 
                        Error Trace:    /workspaces/ibc-go/modules/apps/transfer/keeper/grpc_query_test.go:287
                                                                /go/pkg/mod/github.com/stretchr/testify@v1.9.0/suite/suite.go:115
                        Error:          An error is expected but got nil.
                        Test:           TestKeeperTestSuite/TestEscrowAddress/Case_failure_-_channel_not_found_due_to_invalid_portID
        --- FAIL: TestKeeperTestSuite/TestEscrowAddress/Case_failure_-_channel_not_found_due_to_invalid_channelID (0.43s)
            grpc_query_test.go:287: 
                        Error Trace:    /workspaces/ibc-go/modules/apps/transfer/keeper/grpc_query_test.go:287
                                                                /go/pkg/mod/github.com/stretchr/testify@v1.9.0/suite/suite.go:115
                        Error:          An error is expected but got nil.
                        Test:           TestKeeperTestSuite/TestEscrowAddress/Case_failure_-_channel_not_found_due_to_invalid_channelID
```

which leads me to think that we are testing for the correct path here?

Comment (bznein) [reply modules/apps/transfer/keeper/grpc_query_test.go:268]: (that being said, this means that we can -and should- probably add a test for `validateGRPCRequest`, though maybe this can be delayed until we maybe consolidate it into a single location, as discussed here https://github.com/cosmos/ibc-go/pull/6059#discussion_r1539199138)

File: modules/apps/transfer/keeper/grpc_query_test.go (hunk 9 above)
Comment (bznein) [reply modules/apps/transfer/keeper/grpc_query_test.go:249]: I think this is what we cover in the discussion here https://github.com/cosmos/ibc-go/pull/6059#discussion_r1539724255 :)

File: modules/apps/transfer/keeper/grpc_query_test.go (hunk 8 above)
Comment (crodriguezvega) [reply modules/apps/transfer/keeper/grpc_query_test.go:268]: Oh, yes, my bad. I misinterpreted the function :)

Yes, sounds good to wait to see where that function lands and test it once there.

Comment (crodriguezvega) [reply modules/apps/transfer/keeper/grpc_query_test.go:268]: Although to come back to [Jim's point](https://github.com/cosmos/ibc-go/pull/6059#discussion_r1539215613), maybe it is worth to add a simple test case for both port ID empty and channel ID empty.

Comment (bznein) [reply modules/apps/transfer/keeper/grpc_query_test.go:278]: Sounds good! I removed one of the two "channel not found" test because they were actually both just testing the same line (`hasChannel() == false`) and added one for an empty portID and one for an empty channelID

Comment (crodriguezvega) [reply modules/apps/transfer/keeper/grpc_query_test.go:278]: Great, thank you!

**Review by crodriguezvega — APPROVED** [review]
Great work, @bznein. Looking forward to the follow-up PRs to add the checks in the other gRPCs. :)
